### ISSUE TITLE
hiccup linter

### ISSUE THREAD

**Issue by borkdude** [clj-kondo/clj-kondo#2660]
Lint hiccup problems:

- attribute map in wrong position: `[:div "Hello" {:id "dude"}]`
- function calls (in reagent): `[foo]` foo called with wrong number of arguments
- invalid HTML nestings: `[:span [:div ...]]`

How to trigger hiccup detection:

- Only vector literals in return position of functions
- Only vector literals that start with keyword or symbol that refers to var (not a local, although in reagent you can have local components...?)

This needs research to see how often false positives are triggered.

The trigger could be refined by looking at which libraries are included in the ns form.

**Comment by teodorlu** [comment]
I would love better static analysis for hiccup! Maybe even completion in a future far, far away. Kondo-powered autocomplete for HTML attributes would be amazing.

> vector literals that start with keyword or symbol that refers to var (not a local, although in reagent you can have local components...?)

I'd propose being able to choose "which Hiccup" explicitly. Something like

```clojure
(defmacro h
  "Mark hiccup for static analysis. Options configure static analysis behavior."
  ([hiccup] hiccup)
  ([opts hiccup] hiccup))
```

Either options could control hiccup dialect, strictness and such, or "static analysis stuff" could be related to the qualified macro symbol.

**Comment by borkdude** [comment]
The only problem is that people don't usually use such a macro. That macro has to exist in the various hiccup libraries and then they need to be hardcoded into clj-kondo to be recognized.

**Comment by teodorlu** [comment]
> That macro has to exist in the various hiccup libraries and then they need to be hardcoded into clj-kondo to be recognized.

I was thinking the qualified macro symbol could be a Kondo config option?

Then I could choose to "opt into" Kondo's hiccup niceness by using that macro. And I might just refer to a macro in my own code.

———

Macro call will give a higher barrier for usage. It's an extra wrapping form. Maybe "infer hiccup" is a better place to start for exactly this reason.

If the "guess what is hiccup?" route is taken, the linting would have to be more conservative to avoid false positives.

**Comment by borkdude** [comment]
> I was thinking the qualified macro symbol could be a Kondo config option?

This could work yes

**Comment by borkdude** [comment]
Branch `hiccup` started. Right now I'm checking if a hiccup context is started by checking the keyword name in the hiccup vector.

**Comment by borkdude** [comment]
I'm checking the keyword in the hiccup vector to see if it's an HTML element. However, I'm getting a false positive with:

```
[:map [:success-handler {:optional true} [:=> [:cat :any :double :string] :any]
       :err-handler {:optional true} [:=> [:cat [:fn ::whatever]] :any]
       :pool-size {:optional true} number?
       :max-batch-messages {:optional true} number?
       :max-next-ms {:optional true} number?]]
```

which is something related to malli. `:map` is a valid HTML element name so it interprets `{:optional true}` as an attribute map in the wrong location.

**Comment by borkdude** [comment]
Another similar one:

```
[:field (mt/id :venues :id) {:temporal-unit :month}]
```
